pages/blog/react-as-generator-engine.md

```diff
@@ -0,0 +1,218 @@
+
+Using helper functions from third party packages, in Nunjucks you must apply them as [filters](https://github.com/asyncapi/generator/blob/master/docs/authoring.md#filters). For example, you want to use one function from [`Lodash`](https://lodash.com/) library like `lowerCase`. To do this, you must create a function inside `filters` folder to convert the function to Nunjucks's filter:
+
+```js
+// filters/lowerCase.js
+const _ = require('lodash');  
+const filter = module.exports;
+filter.lowerCase = _.lowerCase;
+```
+
+And then you can use this function inside your template/macro:
+
+```njk
+{{ AsyncAPI rocks! | lowerCase }}
+```
+
+The main problem in this solution is that it creates unnecessary boilerplate - you must create function in the separate file. Another problem is that you operate on name of this helper function which means you must always remember which filters you have included in your template.
+
+In opposite, in React you can use `Lodash` directly in your template:
+
+```js
+import _ from 'lodash';  
+
+function MyComponent() {
+  return _.lowerCase(`AsyncAPI rocks!`);
+}
+```
+
+It is worth mentioning that when using packages in this way, you always operate on the reference to the function, not on its name, so you know what functions you have in the file's scope.
+
+## Cons & Pros
+
+Like any solution, React has its advantages as well as disadvantages.
+
+We can list the advantages:
+
+- Using React, you directly use JS. You don't need to learn custom syntax, only how React works under the hood. This is one of the problems with template engines like Nunjucks.
+- It enables the possiblity of debugging your template (this is not possible with Nunjucks).
+- It provides better error stack traces.
+- Better tool support for development. As we mentioned, writing templates in JS, you use reference to functions/variables, and therefore your IDE can tell you what you can use in given scope.
+- Provides better support for separating code into more manageable chunks/components. You don't need to create `partials` folder (like in Nunjucks) and put every partials inside it. You can create React component whenever you want, also next to the template's source code.
+- Introduces testability of components which is very difficult with Nunjucks. You can split template to separate chunks and test them in separate test cases.
+
+However, it does have disadvantages:
+
+- Common pain when writing templates with React, will be writing indentations and new lines. We know about this problem and in the next integrations of the React in the Generator we will try to minimize this problem as much as possible. However, we have a several helpers in `@asyncapi/generator-react-sdk` package to make the life easier, like [`withIndendation`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withIndendation.ts#L13) or [`withNewLines`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withNewLines.ts#L8).
```

Comment:
```suggestion
- Common pain when writing templates with React is related to indentations and new lines. However, we have several helpers in SDK to make your life easier, like [`withIndendation`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withIndendation.ts#L13) or [`withNewLines`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withNewLines.ts#L8).
```

should we really link to code here `like [`withIndendation`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withIndendation.ts#L13) or [`withNewLines`](https://github.com/asyncapi/generator-react-sdk/blob/master/src/utils/withNewLines.ts#L8)` instead of maybe talking about `Indent` component and `IndentationTypes` ?

Reply:
`Indent` component you use when you operate on JSX tree, `withIndentation` when you operate on pure string.